### PR TITLE
CI

### LINKED ISSUE #35: CI for deferred typed holes
- Create an additional flake output that builds with `-fdefer-typed-holes`
- Configure Hercules to build both

This will give us an additional green tick for `defer-typed-holes`.

https://docs.hercules-ci.com/hercules-ci-agent/evaluation#attributes-herculesCI.onPush
https://mlabs-corp.slack.com/archives/C0290MGE23D/p1676909814213239?thread_ts=1676905274.718879&cid=C0290MGE23D

### PR DESCRIPTION
**Pull request by Geometer1729**
closes #35

### CODE REVIEW

**Comment by AriFordsham** [comment]
I'd like a review from @brainrake

File: flake.nix [hunk 1]
```diff
@@ -38,6 +39,19 @@
             "plutus-simple-model"
           ];
         })
+        ({
+          flake.config.herculesCI = {
+            onPush = {
+              mainChecks.outputs.mainCheck =
+                self.packages.x86_64-linux."hedgehog-plutus-simple:lib:hedgehog-plutus-simple".override
```
Comment (brainrake) [inline flake.nix]: i would move this to a flake output so it is possible to build with `nix build`, and reference it from there

**Review by brainrake — CHANGES_REQUESTED** [review]

File: flake.nix [hunk 2]
```diff
@@ -53,5 +51,10 @@
           };
         })
       ];
+      perSystem = { self', ... }: {
+        packages.hps-main =
```
Comment (AriFordsham) [inline flake.nix]: Can we have a more descriptive name?

Comment (Geometer1729) [reply flake.nix]: Sure. Does `hps-production-flags` sound good? production isn't quite accurate but strict would be confusing.

**Comment by AriFordsham** [comment]
This seems overly complicated. Can you not just set two jobs to two outputs, one with the flag set?

**Comment by Geometer1729** [comment]
> This seems overly complicated. Can you not just set two jobs to two outputs, one with the flag set?

I agree that it's more complicated than I'd like now. I'm not sure what you mean by "set two jobs to two outputs". There are a bunch of other jobs in the default onPush so to keep all those afaict they either need to be enumerated or taken from `self.outputs`
self.outputs contains all the hps-production-flags stuff now so if we want to derive it from self.outputs that has to be filtered out.

There might be a cleaner way to do that filter in nix, I don't know a ton of nix I just looked at the builtins and mapAttrs was the only way I could see to do it. Or we could enumerate the jobs but I really don't like that because when we add stuff it won't show up in ci unless we remember to add it here too.

**Comment by brainrake** [comment]
add the test that you don't always want to run to a non-standard flake output like `extraTestPackages` and only build/run that in the extra job

**Comment by Geometer1729** [comment]
Changing packages to extraTestPackages somehow causes `error: infinite recursion encountered` @brainrake any idea what's causing that or how to fix it?

**Comment by AriFordsham** [comment]
<details>
  <summary>Error log</summary>
  ```
  error: infinite recursion encountered

       at /nix/store/509ign2dbdbqhn7k60bn4wxf4x282nlg-source/lib.nix:74:32:

           73|       lib.evalModules {
           74|         specialArgs = { inherit self flake-parts-lib; inherit (self) inputs; } // specialArgs;
             |                                ^
           75|         modules = [ ./all-modules.nix module ];

       … while evaluating the attribute 'self'

       at /nix/store/509ign2dbdbqhn7k60bn4wxf4x282nlg-source/lib.nix:74:32:

           73|       lib.evalModules {
           74|         specialArgs = { inherit self flake-parts-lib; inherit (self) inputs; } // specialArgs;
             |                                ^
           75|         modules = [ ./all-modules.nix module ];

       … while evaluating the module argument `self' in "/nix/store/509ign2dbdbqhn7k60bn4wxf4x282nlg-source/modules/perSystem.nix":

       … while evaluating anonymous lambda

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/modules.nix:476:44:

          475|       context = name: ''while evaluating the module argument `${name}' in "${key}":'';
          476|       extraArgs = builtins.mapAttrs (name: _:
             |                                            ^
          477|         builtins.addErrorContext (context name)

       … from call site

       … while evaluating 'optionalAttrs'

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/attrsets.nix:609:5:

          608|     # The attribute set to return if `cond` is `true`.
          609|     as:
             |     ^
          610|     if cond then as else {};

       … from call site

       at /nix/store/509ign2dbdbqhn7k60bn4wxf4x282nlg-source/modules/packages.nix:53:7:

           52|     perInput = system: flake:
           53|       optionalAttrs (flake?packages.${system}) {
             |       ^
           54|         packages = flake.packages.${system};

       … while evaluating 'perInput'

       at /nix/store/509ign2dbdbqhn7k60bn4wxf4x282nlg-source/modules/packages.nix:52:24:

           51|
           52|     perInput = system: flake:
             |                        ^
           53|       optionalAttrs (flake?packages.${system}) {

       … from call site

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/types.nix:621:114:

          620|       merge = loc: defs:
          621|         fnArgs: (mergeDefinitions (loc ++ [ "<function body>" ]) elemType (map (fn: { inherit (fn) file; value = fn.value fnArgs; }) defs)).mergedValue;
             |                                                                                                                  ^
          622|       getSubOptions = prefix: elemType.getSubOptions (prefix ++ [ "<function body>" ]);

       … while evaluating the attribute 'value'

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/types.nix:621:106:

          620|       merge = loc: defs:
          621|         fnArgs: (mergeDefinitions (loc ++ [ "<function body>" ]) elemType (map (fn: { inherit (fn) file; value = fn.value fnArgs; }) defs)).mergedValue;
             |                                                                                                          ^
          622|       getSubOptions = prefix: elemType.getSubOptions (prefix ++ [ "<function body>" ]);

       … while evaluating 'dischargeProperties'

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/modules.nix:810:25:

          809|   */
          810|   dischargeProperties = def:
             |                         ^
          811|     if def._type or "" == "merge" then

       … from call site

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/modules.nix:739:137:

          738|         defs' = concatMap (m:
          739|           map (value: { inherit (m) file; inherit value; }) (builtins.addErrorContext "while evaluating definitions from `${m.file}':" (dischargeProperties m.value))
             |                                                                                                                                         ^
          740|         ) defs;

       … while evaluating definitions from `/nix/store/509ign2dbdbqhn7k60bn4wxf4x282nlg-source/modules/packages.nix':

       … while evaluating anonymous lambda

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/modules.nix:738:28:

          737|         # Process mkMerge and mkIf properties.
          738|         defs' = concatMap (m:
             |                            ^
          739|           map (value: { inherit (m) file; inherit value; }) (builtins.addErrorContext "while evaluating definitions from `${m.file}':" (dischargeProperties m.value))

       … from call site

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/modules.nix:738:17:

          737|         # Process mkMerge and mkIf properties.
          738|         defs' = concatMap (m:
             |                 ^
          739|           map (value: { inherit (m) file; inherit value; }) (builtins.addErrorContext "while evaluating definitions from `${m.file}':" (dischargeProperties m.value))

       … while evaluating the attribute 'values'

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/modules.nix:851:7:

          850|     in {
          851|       values = concatMap (def: if getPrio def == highestPrio then [(strip def)] else []) defs;
             |       ^
          852|       inherit highestPrio;

       … while evaluating the attribute 'values'

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/modules.nix:752:9:

          751|       in {
          752|         values = defs''';
             |         ^
          753|         inherit (defs'') highestPrio;

       … while evaluating the attribute 'mergedValue'

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/modules.nix:758:5:

          757|     # Type-check the remaining definitions, and merge them. Or throw if no definitions.
          758|     mergedValue =
             |     ^
          759|       if isDefined then

       … while evaluating 'merge'

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/types.nix:621:9:

          620|       merge = loc: defs:
          621|         fnArgs: (mergeDefinitions (loc ++ [ "<function body>" ]) elemType (map (fn: { inherit (fn) file; value = fn.value fnArgs; }) defs)).mergedValue;
             |         ^
          622|       getSubOptions = prefix: elemType.getSubOptions (prefix ++ [ "<function body>" ]);

       … from call site

       at /nix/store/509ign2dbdbqhn7k60bn4wxf4x282nlg-source/modules/perSystem.nix:38:32:

           37|           _module.args.inputs' = mapAttrs (k: rootConfig.perInput system) self.inputs;
           38|           _module.args.self' = rootConfig.perInput system self;
             |                                ^
           39|         };

       … while evaluating the attribute 'value'

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/types.nix:552:58:

          551|         # Push down position info.
          552|         (map (def: mapAttrs (n: v: { inherit (def) file; value = v; }) def.value) defs);
             |                                                          ^
          553|       emptyValue = { value = {}; };

       … while evaluating 'dischargeProperties'

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/modules.nix:810:25:

          809|   */
          810|   dischargeProperties = def:
             |                         ^
          811|     if def._type or "" == "merge" then

       … from call site

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/modules.nix:739:137:

          738|         defs' = concatMap (m:
          739|           map (value: { inherit (m) file; inherit value; }) (builtins.addErrorContext "while evaluating definitions from `${m.file}':" (dischargeProperties m.value))
             |                                                                                                                                         ^
          740|         ) defs;

       … while evaluating definitions from `/nix/store/509ign2dbdbqhn7k60bn4wxf4x282nlg-source/modules/perSystem.nix':

       … while evaluating anonymous lambda

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/modules.nix:738:28:

          737|         # Process mkMerge and mkIf properties.
          738|         defs' = concatMap (m:
             |                            ^
          739|           map (value: { inherit (m) file; inherit value; }) (builtins.addErrorContext "while evaluating definitions from `${m.file}':" (dischargeProperties m.value))

       … from call site

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/modules.nix:738:17:

          737|         # Process mkMerge and mkIf properties.
          738|         defs' = concatMap (m:
             |                 ^
          739|           map (value: { inherit (m) file; inherit value; }) (builtins.addErrorContext "while evaluating definitions from `${m.file}':" (dischargeProperties m.value))

       … while evaluating the attribute 'values'

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/modules.nix:851:7:

          850|     in {
          851|       values = concatMap (def: if getPrio def == highestPrio then [(strip def)] else []) defs;
             |       ^
          852|       inherit highestPrio;

       … while evaluating the attribute 'values'

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/modules.nix:752:9:

          751|       in {
          752|         values = defs''';
             |         ^
          753|         inherit (defs'') highestPrio;

       … while evaluating the attribute 'optionalValue.value'

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/modules.nix:770:5:

          769|
          770|     optionalValue =
             |     ^
          771|       if isDefined then { value = mergedValue; }

       … while evaluating anonymous lambda

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/types.nix:546:29:

          545|       merge = loc: defs:
          546|         zipAttrsWith (name: defs:
             |                             ^
          547|           let merged = mergeDefinitions (loc ++ [name]) elemType defs;

       … from call site

       … while evaluating the module argument `self'' in ":anon-12:anon-1":

       … while evaluating anonymous lambda

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/modules.nix:476:44:

          475|       context = name: ''while evaluating the module argument `${name}' in "${key}":'';
          476|       extraArgs = builtins.mapAttrs (name: _:
             |                                            ^
          477|         builtins.addErrorContext (context name)

       … from call site

       … while evaluating 'id'

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/trivial.nix:14:5:

           13|     # The value to return
           14|     x: x;
             |     ^
           15|

       … from call site

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/generators.nix:272:47:

          271|           let
          272|             evalNext = x: mapAny (depth + 1) (transform (depth + 1) x);
             |                                               ^
          273|           in

       … while evaluating anonymous lambda

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/generators.nix:270:40:

          269|           else id;
          270|         mapAny = with builtins; depth: v:
             |                                        ^
          271|           let

       … from call site

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/generators.nix:272:27:

          271|           let
          272|             evalNext = x: mapAny (depth + 1) (transform (depth + 1) x);
             |                           ^
          273|           in

       … while evaluating 'evalNext'

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/generators.nix:272:24:

          271|           let
          272|             evalNext = x: mapAny (depth + 1) (transform (depth + 1) x);
             |                        ^
          273|           in

       … from call site

       … while evaluating 'go'

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/generators.nix:295:18:

          294|     let
          295|     go = indent: v: with builtins;
             |                  ^
          296|     let     isPath   = v: typeOf v == "path";

       … from call site

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/generators.nix:345:57:

          344|               (name: value:
          345|                 "${libStr.escapeNixIdentifier name} = ${go (indent + "  ") value};") v)
             |                                                         ^
          346|         + outroSpace + "}"

       … while evaluating anonymous lambda

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/generators.nix:344:22:

          343|           + libStr.concatStringsSep introSpace (libAttr.mapAttrsToList
          344|               (name: value:
             |                      ^
          345|                 "${libStr.escapeNixIdentifier name} = ${go (indent + "  ") value};") v)

       … from call site

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/attrsets.nix:478:16:

          477|     attrs:
          478|     map (name: f name attrs.${name}) (attrNames attrs);
             |                ^
          479|

       … while evaluating anonymous lambda

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/attrsets.nix:478:10:

          477|     attrs:
          478|     map (name: f name attrs.${name}) (attrNames attrs);
             |          ^
          479|

       … from call site

       … while evaluating 'go'

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/generators.nix:295:18:

          294|     let
          295|     go = indent: v: with builtins;
             |                  ^
          296|     let     isPath   = v: typeOf v == "path";

       … from call site

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/options.nix:355:10:

          354|       prettyEval = builtins.tryEval
          355|         (lib.generators.toPretty { }
             |          ^
          356|           (lib.generators.withRecursion { depthLimit = 10; throwOnDepthLimit = false; } def.value));

       … while evaluating anonymous lambda

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/options.nix:351:38:

          350|
          351|   showDefs = defs: concatMapStrings (def:
             |                                      ^
          352|     let

       … from call site

       … while evaluating 'concatMapStrings'

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/strings.nix:54:25:

           53|   */
           54|   concatMapStrings = f: list: concatStrings (map f list);
             |                         ^
           55|

       … from call site

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/options.nix:351:20:

          350|
          351|   showDefs = defs: concatMapStrings (def:
             |                    ^
          352|     let

       … while evaluating 'showDefs'

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/options.nix:351:14:

          350|
          351|   showDefs = defs: concatMapStrings (def:
             |              ^
          352|     let

       … from call site

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/modules.nix:295:25:

          294|                       "while evaluating a definition from `${firstDef.file}'"
          295|                       ( showDefs [ firstDef ])
             |                         ^
          296|                     );

       … while evaluating a definition from `<unknown-file>, via option perSystem'

       … while evaluating the error message for definitions for `perSystem.aarch64-darwin.extraTestPackages', which is an option that does not exist

       … while evaluating the attribute 'options'

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/modules.nix:340:9:

          339|       result = withWarnings {
          340|         options = checked options;
             |         ^
          341|         config = checked (removeAttrs config [ "_module" ]);

       … while evaluating 'apply'

       at /nix/store/509ign2dbdbqhn7k60bn4wxf4x282nlg-source/modules/perSystem.nix:41:24:

           40|       });
           41|       apply = modules: system:
             |                        ^
           42|         let

       … from call site

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/attrsets.nix:552:43:

          551|     f:
          552|     listToAttrs (map (n: nameValuePair n (f n)) names);
             |                                           ^
          553|

       … while evaluating the attribute 'aarch64-darwin'

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/attrsets.nix:423:14:

          422|     value:
          423|     { inherit name value; };
             |              ^
          424|

       … while evaluating anonymous lambda

       at /nix/store/509ign2dbdbqhn7k60bn4wxf4x282nlg-source/modules/apps.nix:89:17:

           88|           (filterAttrs
           89|             (k: v: v ? apps)
             |                 ^
           90|             config.allSystems

       … from call site

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/attrsets.nix:306:62:

          305|     set:
          306|     listToAttrs (concatMap (name: let v = set.${name}; in if pred name v then [(nameValuePair name v)] else []) (attrNames set));
             |                                                              ^
          307|

       … while evaluating anonymous lambda

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/attrsets.nix:306:29:

          305|     set:
          306|     listToAttrs (concatMap (name: let v = set.${name}; in if pred name v then [(nameValuePair name v)] else []) (attrNames set));
             |                             ^
          307|

       … from call site

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/attrsets.nix:306:18:

          305|     set:
          306|     listToAttrs (concatMap (name: let v = set.${name}; in if pred name v then [(nameValuePair name v)] else []) (attrNames set));
             |                  ^
          307|

       … while evaluating 'filterAttrs'

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/attrsets.nix:305:5:

          304|     # The attribute set to filter
          305|     set:
             |     ^
          306|     listToAttrs (concatMap (name: let v = set.${name}; in if pred name v then [(nameValuePair name v)] else []) (attrNames set));

       … from call site

       at /nix/store/509ign2dbdbqhn7k60bn4wxf4x282nlg-source/modules/apps.nix:88:12:

           87|           (k: v: v.apps)
           88|           (filterAttrs
             |            ^
           89|             (k: v: v ? apps)

       … while evaluating the attribute 'condition'

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/modules.nix:881:14:

          880|     { _type = "if";
          881|       inherit condition content;
             |              ^
          882|     };

       … while evaluating 'dischargeProperties'

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/modules.nix:810:25:

          809|   */
          810|   dischargeProperties = def:
             |                         ^
          811|     if def._type or "" == "merge" then

       … from call site

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/modules.nix:739:137:

          738|         defs' = concatMap (m:
          739|           map (value: { inherit (m) file; inherit value; }) (builtins.addErrorContext "while evaluating definitions from `${m.file}':" (dischargeProperties m.value))
             |                                                                                                                                         ^
          740|         ) defs;

       … while evaluating definitions from `/nix/store/509ign2dbdbqhn7k60bn4wxf4x282nlg-source/modules/apps.nix':

       … while evaluating anonymous lambda

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/modules.nix:738:28:

          737|         # Process mkMerge and mkIf properties.
          738|         defs' = concatMap (m:
             |                            ^
          739|           map (value: { inherit (m) file; inherit value; }) (builtins.addErrorContext "while evaluating definitions from `${m.file}':" (dischargeProperties m.value))

       … from call site

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/modules.nix:738:17:

          737|         # Process mkMerge and mkIf properties.
          738|         defs' = concatMap (m:
             |                 ^
          739|           map (value: { inherit (m) file; inherit value; }) (builtins.addErrorContext "while evaluating definitions from `${m.file}':" (dischargeProperties m.value))

       … while evaluating the attribute 'values'

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/modules.nix:851:7:

          850|     in {
          851|       values = concatMap (def: if getPrio def == highestPrio then [(strip def)] else []) defs;
             |       ^
          852|       inherit highestPrio;

       … while evaluating the attribute 'values'

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/modules.nix:752:9:

          751|       in {
          752|         values = defs''';
             |         ^
          753|         inherit (defs'') highestPrio;

       … while evaluating the attribute 'mergedValue'

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/modules.nix:758:5:

          757|     # Type-check the remaining definitions, and merge them. Or throw if no definitions.
          758|     mergedValue =
             |     ^
          759|       if isDefined then

       … while evaluating the option `flake.apps':

       … while evaluating the attribute 'value'

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/modules.nix:723:9:

          722|     in warnDeprecation opt //
          723|       { value = builtins.addErrorContext "while evaluating the option `${showOption loc}':" value;
             |         ^
          724|         inherit (res.defsFinal') highestPrio;

       … while evaluating anonymous lambda

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/modules.nix:265:72:

          264|           # For definitions that have an associated option
          265|           declaredConfig = mapAttrsRecursiveCond (v: ! isOption v) (_: v: v.value) options;
             |                                                                        ^
          266|

       … from call site

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/attrsets.nix:532:20:

          531|               then recurse (path ++ [name]) value
          532|               else f (path ++ [name]) value;
             |                    ^
          533|         in mapAttrs g;

       … while evaluating 'g'

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/attrsets.nix:529:19:

          528|           g =
          529|             name: value:
             |                   ^
          530|             if isAttrs value && cond value

       … from call site

       … while evaluating anonymous lambda

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/attrsets.nix:710:24:

          709|     let f = attrPath:
          710|       zipAttrsWith (n: values:
             |                        ^
          711|         let here = attrPath ++ [n]; in

       … from call site

       … while evaluating anonymous lambda

       at /nix/store/509ign2dbdbqhn7k60bn4wxf4x282nlg-source/lib.nix:80:23:

           79|       # filter away top-level nulls
           80|       filterAttrs (k: v: v != null) (flake-parts-lib.evalFlakeModule args module).config.flake;
             |                       ^
           81|

       … from call site

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/attrsets.nix:306:62:

          305|     set:
          306|     listToAttrs (concatMap (name: let v = set.${name}; in if pred name v then [(nameValuePair name v)] else []) (attrNames set));
             |                                                              ^
          307|

       … while evaluating anonymous lambda

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/attrsets.nix:306:29:

          305|     set:
          306|     listToAttrs (concatMap (name: let v = set.${name}; in if pred name v then [(nameValuePair name v)] else []) (attrNames set));
             |                             ^
          307|

       … from call site

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/attrsets.nix:306:18:

          305|     set:
          306|     listToAttrs (concatMap (name: let v = set.${name}; in if pred name v then [(nameValuePair name v)] else []) (attrNames set));
             |                  ^
          307|

       … while evaluating 'filterAttrs'

       at /nix/store/3qi3vc8wq3lxndfi30pmva975j5n43hw-source/lib/attrsets.nix:305:5:

          304|     # The attribute set to filter
          305|     set:
             |     ^
          306|     listToAttrs (concatMap (name: let v = set.${name}; in if pred name v then [(nameValuePair name v)] else []) (attrNames set));

       … from call site

       at /nix/store/509ign2dbdbqhn7k60bn4wxf4x282nlg-source/lib.nix:80:7:

           79|       # filter away top-level nulls
           80|       filterAttrs (k: v: v != null) (flake-parts-lib.evalFlakeModule args module).config.flake;
             |       ^
           81|

       … while evaluating 'mkFlake'

       at /nix/store/509ign2dbdbqhn7k60bn4wxf4x282nlg-source/lib.nix:78:21:

           77|
           78|     mkFlake = args: module:
             |                     ^
           79|       # filter away top-level nulls

       … from call site

       at /nix/store/19xsx9fccaic8xyvlq58l304bhdnvrr2-source/flake.nix:19:75:

           18|
           19|   outputs = inputs@{ self, tooling, plutus-simple-model, plutarch, ... }: tooling.lib.mkFlake
             |                                                                           ^
           20|     { inherit self; }

       … while evaluating 'outputs'

       at /nix/store/19xsx9fccaic8xyvlq58l304bhdnvrr2-source/flake.nix:19:13:

           18|
           19|   outputs = inputs@{ self, tooling, plutus-simple-model, plutarch, ... }: tooling.lib.mkFlake
             |             ^
           20|     { inherit self; }

       … from call site

       at «string»:45:21:

           44|
           45|           outputs = flake.outputs (inputs // { self = result; });
             |                     ^
           46|

       … while evaluating anonymous lambda

       at «string»:10:13:

            9|     builtins.mapAttrs
           10|       (key: node:
             |             ^
           11|         let

       … from call site

       … while evaluating anonymous lambda

       at «string»:2:23:

            1|
            2| lockFileStr: rootSrc: rootSubdir:
             |                       ^
            3|

       … from call site

  ```
</details>

**Comment by Geometer1729** [comment]
Go back to filter solution as per slack conversation

**Review by AriFordsham — APPROVED** [review]
Shame about the CI not passing currently, but I think the CI is sound.

**Pull request merged**